projects/noloan/app/filter/bayes/BayesFilter.java

```diff
@@ -0,0 +1,168 @@
+package projects.noloan.app.filter.bayes;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import projects.noloan.app.filter.SpamFilterAlgorithm;
+import projects.noloan.app.filter.bayes.Protos.*;
+
+public class BayesFilter implements SpamFilterAlgorithm {
+  private HashMap<String, WordInfo> words = new HashMap<>();
+  private BayesianProbabilityCalculationStrategy strategy;
+  private int totalSpamCount = 0;
+  private int totalHamCount = 0;
+
+  interface BayesianProbabilityCalculationStrategy {
+    double probability(List<Protos.WordInfo> message);
+  }
+
+  static class SimpleProbabilityStrategy implements BayesianProbabilityCalculationStrategy {
+
+    @Override
+    public double probability(List<WordInfo> message) {
+      double positive_probabilities_product = 1.0f;
+      double negative_probabilities_product = 1.0f;
+      for (int i = 0; i < message.size(); i++) {
+        WordInfo wordInfo = message.get(i);
+        positive_probabilities_product *= wordInfo.getSpamProbability();
+        negative_probabilities_product *= (1.0f - wordInfo.getSpamProbability());
+      }
+      double probabilities_products_sum = positive_probabilities_product + negative_probabilities_product;
+      return positive_probabilities_product / probabilities_products_sum;
+    }
+  }
+
+  static class LogProbabilityStrategy implements BayesianProbabilityCalculationStrategy {
```

Comment:
Let's choose what we think is the best implementation - simple or log, and make is just a method in BayesFilter.